libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -807,15 +807,31 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
 	{
 		break;
 	}
-	case FunctionType::Kind::GasLeft:
+	case FunctionType::Kind::AddMod:
+	case FunctionType::Kind::MulMod:
 	{
-		define(_functionCall) << "gas()\n";
-		break;
+		solAssert(arguments.size() == 3 && parameterTypes.size() == 3, "");
+		Whiskers templ("if iszero(<var>) { revert(0, 0) }\n");
+		templ("var", expressionAsType(*arguments[2], *(parameterTypes[2])));
+		m_code << templ.render();
+		[[fallthrough]];
 	}
+	case FunctionType::Kind::GasLeft:
 	case FunctionType::Kind::Selfdestruct:
+	case FunctionType::Kind::BlockHash:
 	{
-		solAssert(arguments.size() == 1, "");
-		define(_functionCall) << "selfdestruct(" << expressionAsType(*arguments.front(), *parameterTypes.front()) << ")\n";
+		static map<FunctionType::Kind, string> functions = {
+			{FunctionType::Kind::AddMod, "addmod"},
+			{FunctionType::Kind::MulMod, "mulmod"},
+			{FunctionType::Kind::GasLeft, "gas"},
+			{FunctionType::Kind::Selfdestruct, "selfdestruct"},
+			{FunctionType::Kind::BlockHash, "blockhash"},
+		};
+		solAssert(functions.find(functionType->kind()) != functions.end(), "");
+		string args;
+		for (size_t i = 0; i < arguments.size(); ++i)
+			args += (args.empty() ? "" : ", ") + expressionAsType(*arguments.at(i), *(parameterTypes.at(i)));
```

Comment:
Some other parts of the code use constructs like this:
```
			if (functionType->takesArbitraryParameters())
				args.emplace_back(IRVariable(*arguments[i]).commaSeparatedList());
			else	
				args.emplace_back(convert(*arguments[i], *parameterTypes[i]).commaSeparatedList());
```

Would something similar work here?

Reply:
`expressionAsType` calls `convert`.

Reply:
This pattern then:
```
		if (IRVariable(*arg).type().sizeOnStack() > 0)
			argumentStrings.emplace_back(IRVariable(*arg).commaSeparatedList());
	}		}
	string argumentString = argumentStrings.empty() ? ""s : (", " + joinHumanReadable(argumentStrings));		string argumentString = argumentStrings.empty() ? ""s : (", " + joinHumanReadable(argumentStrings));
```

Basically using `joinHumanReadable` instead of manual concatenation.

Reply:
We need a version of `joinHumanReadable` that does not add a comma when the element is empty. Then the code you cited could be very much simplified.